src/compose/application-manager.ts

```diff
@@ -694,7 +702,7 @@ function saveAndRemoveImages(
 			const isNotSaved = !_.some(availableWithoutIds, (img) =>
 				_.isEqual(img, targetImage),
 			);
-			return isActuallyAvailable && isNotSaved;
+			return isActuallyAvailable && isNotSaved && !isPurgingVolumes;
```

Comment:
I'm not sure I would modify this here @cywang117,  I would prefer to prevent the application manager from performing the [image cleanup step](https://github.com/balena-os/balena-supervisor/blob/master/src/compose/application-manager.ts#L209) on any intermediate state. The image database is going away and when that happens, the cleanup logic will be done through a `image prune` step, and it will still make sense to make it here.

Even if/when we move to actors, we will probably want to notify the application manager actor to skip cleanup on intermediate states, so that this logic will be maintained.

Reply:
@pipex After checking out the old implementation, no cleanup step is ever generated. The doPurge steps are 100% as listed above: kill --> noop --> removeImage --> removeVolume --> fetch --> createVolume + start. We never enter the `if (cleanupNeeded)` block. This is because currently the cleanupNeeded boolean is determined by whether the docker image ID is in the image table, which will definitely change once the image table is removed.

From what I've investigated, the error is entirely in [`saveAndRemoveImages`](https://github.com/balena-os/balena-supervisor/blob/4f41901275df971e8769ab31f6d73234d72c1017/src/compose/application-manager.ts#L604), which generates the unnecessary removeImage/saveImage steps based on incomplete logic which hasn't taken volume purging status into account. I see the rationale behind not touching saveAndRemove though, since it'll be removed or changed significantly with image table removal.

If we don't want to change saveAndRemove, I think there'd be two ways to go about fixing this issue:
1. Wrap the [saveAndRemove call in getRequiredSteps](https://github.com/balena-os/balena-supervisor/blob/4f41901275df971e8769ab31f6d73234d72c1017/src/compose/application-manager.ts#L215) with a isPurging conditional, or isApplyingIntermediate conditional so it's skipped when purging.
2. Filter out removeImage / saveImage steps when isPurging / isApplyingIntermediate is true.

I'm more inclined to go with option 1 since it feels less hack-y.

Also, I'm actually more inclined to keep isPurging as a conditional over isApplyingIntermediate, since there could be a case where Supervisor is applying intermediate state but needs a remove or saveImage step (there aren't any of these cases right now, but there *could* be)

What do you think?

Reply:
> Wrap the saveAndRemove call in getRequiredSteps with a isPurging conditional, or isApplyingIntermediate conditional so it's skipped when purging.

This is what I meant, sorry. I would add the `isPurging` or `isIntermediate` to the `if` on line 209 and skip both the cleanup and the `saveAndRemoveImages` call altogether. When we remove the image database, the `saveAndRemoveImages` will go away and we will always perform a image prune as `cleanup` step.

I honestly prefer `isIntermediateTarget` or `isLocalTarget` or something, now we only do it for purging, but we also will need to change the `stop` and `restart` actions to use intermediate targets when we move the composer to its own library. If we do just `isPurging` that opens the floodgates to adding a bunch of other conditions.

Reply:
@pipex Got it! That makes sense, I'll update the PR. I've also added tests so this PR can move out of draft status :)